ocaml/src/gblob_access.ml

```diff
@@ -0,0 +1,350 @@
+          (fun _ -> fn)
+          results
+        >>= fun () ->
+        Prelude.with_timing_lwt
+          (fun () -> Lwt_list.iter_s _f results)
+        >>= fun (took,()) ->
+        let () = AsdStatistics.new_delta _statistics _READ_BATCH_CB took in
+        Lwt.return_unit
+      )
+      (fun () ->
+        List.iter Fragment.free_bytes fragments;
+        Prelude.with_timing_lwt
+          (fun () ->IOExecFile.file_close handle)
+        >>= fun (took,()) ->
+        AsdStatistics.new_delta _statistics _FILE_CLOSE took;
+        Lwt.return_unit
+      )
+
+  method push_blob_data fnr len slices f =
+    let _f (correction,ec) =
+      let (completion_id,off',off,len, len',fragment) = correction in
+      let buffer = Fragment.get_bytes  fragment in
+      let left = off - off' in
+      f (off, len) buffer left
+    in
+    self # _push_blob_data fnr len slices _f
+
+  method send_blob_data_to fnr len slices nfd =
+    let _f (correction, ec) =
+      let (completion_id,off',off, len,len',fragment) = correction in
+      let buffer = Fragment.get_bytes fragment in
+      let left = off' - off in
+      Net_fd.write_all_lwt_bytes nfd buffer left len
+    in
+    self # _push_blob_data fnr len slices _f
+
+
+  method private _inner () : unit Lwt.t =
+    Lwt_log.debug "_inner ()" >>= fun () ->
+    let buf = Lwt_bytes.create 256 in
+    let rec loop () =
+      Lwt_unix.wait_read _reap_fd  >>= fun () ->
+      IOExecFile.reap _reap_fd buf >>= fun (n, ss) ->
+      (*Lwt_log.debug_f "reaped:%i" n >>= fun () -> *)
+      List.iter
+        (fun s ->
+          let completion_id = IOExecFile.get_completion_id s in
+          let ec = IOExecFile.get_error_code s in
+          let awake = Hashtbl.find _outstanding completion_id in
+          Lwt.wakeup awake ec;
+          Hashtbl.remove _outstanding completion_id
+        ) ss;
+      loop ()
+    in
+    loop ()
+
+  method _wait_for_completion completion_id : int32 Lwt.t =
+    let sleep, awake = Lwt.wait () in
+    let () = Hashtbl.add _outstanding completion_id awake in
+    sleep
+
+  method write_blob fnr blob ~post_write ~sync_parent_dirs =
+
+    let dir, _, file_path = self # _get_file_dir_name_path fnr in
+    self # ensure_dir_exists dir ~sync:sync_parent_dirs >>= fun () ->
+
+    let completion_id = self # next_completion_id  in
+    Lwt_log.debug_f "write_blob ~fnr:%Li completion_id:%Li" fnr completion_id >>= fun () ->
+    let len = Blob.length blob in
+    let alignment = 4096 in
+    let remainder = len mod alignment in
+    if remainder <> 0
+    then
+      begin
+        Lwt_log.info_f
+          "ragged blob size: fnr:%Li len:%i (0x%0x) using ordinary write"
+          fnr len len
+        >>= fun () ->
+        Lwt_extra2.with_fd
+          file_path
+          ~flags:Lwt_unix.([ O_WRONLY; O_CREAT; O_EXCL; ])
+          ~perm:0o664
+          (fun fd ->
+            let extra = alignment - remainder in
+            let padded_size = len + extra  in
```

Comment:
IIRC consensus was that this part (padding the blobs written to disk) would no longer be necessary.
And even if it was (or still is) necessary then making the copy would not be needed (just write the unchanged blob first, followed by some padding).